flocker/node/agents/test/test_blockdevice.py

```diff
@@ -5249,3 +5257,231 @@ def test_get_device_path_until_detach(self):
 
         self.assertRaises(UnattachedVolume,
                           self.cache.get_device_path, attached_id1)
+
+
+@attributes(["_proxy_object", "_callback"], apply_with_init=False)
+class CallbackProxy(object):
+    """
+    A proxy to another object that calls a callback before and after each call
+    to the underlying object. This enables tests where you want to inject code
+    before or after any calls to an underlying API.
+
+    Note that this proxy assumes that the underlying object only has public
+    methods and not public attributes.
+
+    :ivar _proxy_object: The object that is being proxied to.
+    :ivar _callback: The callback to be called before and after every call to a
+        method of _proxy_object.
+    """
+
+    def __init__(self, proxy_object, callback):
+        self._proxy_object = proxy_object
+        self._callback = callback
+
+    def __getattr__(self, name):
+        """
+        The implementation of the proxy. Gets the attribute on the underlying
+        object and returns a wrapped function that calls the callback before
+        and after its execution.
+
+        :param name: The name of the method on the proxy object that should be
+            called.
+        """
+        method = getattr(self._proxy_object, name)
+
+        def proxied_function(*args, **kwargs):
+            self._callback('Before %s' % name)
```

Comment:
:+1: Also I realized that stack traces work better than random strings for debugging so I've removed for now. If I'm tempted to add back I'll be sure to use structured data.